README
PURPOSE AND OPERATION:

The module performs the checks if the number of HTTP requests from a single IP 
exceeds a certain allowed maximum during a chosen period of time. If so, it
makes the HTTP server respond to the requests from such IP addresses with a
chosen HTTP response code (e.g, 402). The checks can be performed either 
globally for the whole web-server, or for certain URLs that match some pattern. 

INSTALLATION:
This module works with Apache2 verions 2.0.52 and greater and requires the 
following packages:
apr_memcache http://svn.northnitch.com/apr_memcache/trunk/
(uses the locally included code, and will not install it)

m4 'helper' modules http://svn.outoforder.cc/svn/autoconf/trunk/m4

memcached http://code.sixapart.com/svn/memcached/trunk/


To build the module, run
./configure
make
Then `su' to root and  
make install
Usually, this will do the job. See the output of ./configure --help for more
configuration options.

The sample configuration to load and set up the module is located in 
mod_mod_ip_count.conf. 

This code is licenced under ASF License, version 2.0

autoconf magic was stolen/from (and the original author didn't succeed much in 
that, so many modifications were done and there are many more to do) 
mod_transform.. a excellent module from svn.outoforder.cc 

USAGE AND CONFIGURATION DIRECTIVES
Once installed, the module needs to be configured globally (a per-virtual-host
configuration seems useless now). Here is the list of presently available
configuration directives and their usage:

MemCacheServers
A list of memcached servers the module will use. By default, a newly installed
memcached will bind to localhost:11211, and it is the default setting for the
module, so you may just skip this setting if you didn't reconfigure memcached

UriList
List of sample patterns. If the directive is set, only the requests containing
such pattern inside the URI part will be counted, and the rest will be ignored.
If the directive is omitted the module will operate on all requests. 

MemCacheMaxTime
MemCacheMaxRequests
Set the time in seconds and the number of requests per IP to serve during this
time. The requests that come after such limit is exceeded are replied with some
HTTP code instead of being processed normally. Note that a single user opening
a single webpage generates several requests (as an excelent example, see your
Apache log after directing your web browser to a simple phpinfo page). 
Therefore, you might probably have to gather some per-page request statistics
for your site to figure out the appropriate values.

MemCacheBlockTime
Optionally, set the time in seconds during which the request from IP's which
have exceeded their maximum are being blocked.

HttpResponse
Numeric HTTP return code to be sent in response to the requests after their 
number excceeds the permitted limit. If omitted, the default 402 return code
will be used

MemCacheAllow
IP-adresses and/or `env' variables with which an unlimited number of requests
is permitted

Latency
When it's impossible to tell exatly, how many requests will user generate
with a single hit on a web address, this directive may be used to count
all the requests coming during the specified period of time (in microseconds,
0,000001 second) from a single IP as a single request. The option is not
mandatory


ChangeLog

2008-11-21 sin 2.2.0-cvs
        * Configure script now detects whether to use apxs or apxs2
	* Removed needless check for g++ presence that prevented module from 
	    building on the systems where g++ is not installed

2008-04-16 sin 2.1.1 bugfix release
	* Fixed 'Segmentation Fault' when there are no memcache servers. 

2008-03-19 sin 2.1.0 release
	* Added support for per-url request processing via pattern matching
	* HTTP server response is now customizable through config file
	* A blocking time period can be set for IPs from which the number of 
	    requests exceeds the allowed maximum.
	* Added whitelisting (oops... MemCacheAllow option in the sample config
	    until I discovered it being only there but not in the code :D )
	* Added `Latency` (see README for explanation)
	* Now the included memcache client interface can be compiled on both
	    Apache 2.0.x and Apache 2.2.x
	* Some minor bugfixes and some new bugs, yet unknown. As usual ^)
	


memcache/apr_memcache.h
/* $Id: apr_memcache.h,v 1.2 2008/11/22 09:10:08 proger Exp $ */

/* Copyright 2004 Paul Querna
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef APR_MEMCACHE_H
#define APR_MEMCACHE_H

/**
 * @file apr_memcache.h
 * @brief Client interface for memcached
 * @remark To use this interface you must have a separate memcached
 * server running. See the memcached website at http://www.danga.com/memcached/
 * for more information.
 */

#include "apr.h"
#include "apr_pools.h"
#include "apr_time.h"
#include "apr_strings.h"
#include "apr_network_io.h"
#include "apr_ring.h"
#include "apr_buckets.h"
#include "apr_reslist.h"

#ifdef __cplusplus
extern "C" {
#endif /* __cplusplus */

/**
 * @defgroup APR_Util_MC Memcached Client Routines
 * @ingroup APR_Util
 * @{
 */

/** Specifies the status of a memcached server */
typedef enum
{
    APR_MC_SERVER_LIVE, /**< Server is alive and responding to requests */
    APR_MC_SERVER_DEAD  /**< Server is not responding to requests */
} apr_memcache_server_status_t;

/** Opaque memcache client connection object */
typedef struct apr_memcache_conn_t apr_memcache_conn_t;

/** Memcache Server Info Object */
typedef struct apr_memcache_server_t apr_memcache_server_t;
struct apr_memcache_server_t
{
    const char *host; /**< Hostname of this Server */
    apr_port_t port; /**< Port of this Server */
    apr_memcache_server_status_t status; /**< @see apr_memcache_server_status_t */
#if APR_HAS_THREADS || defined(DOXYGEN)
    apr_reslist_t *conns; /**< Resource list of actual client connections */
#else
    apr_memcache_conn_t *conn;   
#endif
    apr_pool_t *p; /** Pool to use for private allocations */
};

/** Container for a set of memcached servers */
typedef struct
{
    apr_uint32_t flags; /**< Flags, Not currently used */
    apr_uint16_t nalloc; /**< Number of Servers Allocated */
    apr_uint16_t ntotal; /**< Number of Servers Added */
    apr_memcache_server_t **live_servers; /**< Array of Servers */
    apr_pool_t *p; /** Pool to use for allocations */
} apr_memcache_t;

/**
 * Creates a crc32 hash used to split keys between servers
 * @param data Data to be hashed
 * @param data_len Length of the data to use
 * @return crc32 hash of data
 * @remark The crc32 hash is not compatible with old memcached clients.
 */
APR_DECLARE(apr_uint32_t)
apr_memcache_hash(const char* data, apr_uint32_t data_len);

/**
 * Picks a server based on a hash
 * @param mc The memcache client object to use
 * @param hash Hashed value of a Key
 * @return server that controls specified hash
 * @see apr_memcache_hash
 */
APR_DECLARE(apr_memcache_server_t *)
apr_memcache_find_server_hash(apr_memcache_t *mc, const apr_uint32_t hash);

/**
 * Adds a server to a client object
 * @param mc The memcache client object to use
 * @param ms Server to add
 * @remark Adding servers is not thread safe, and should be done once at startup.
 * @warning Changing servers after startup may cause keys to go to
 * different servers.
 */
APR_DECLARE(apr_status_t)
apr_memcache_add_server(apr_memcache_t * mc,
                        apr_memcache_server_t *server);


/**
 * Finds a Server object based on a hostname/port pair
 * @param mc The memcache client object to use
 * @param host Hostname of the server
 * @param port Port of the server
 * @return Server with matching Hostname and Port, or NULL if none was found.
 */
APR_DECLARE(apr_memcache_server_t *)
apr_memcache_find_server(apr_memcache_t *mc,
                         const char *host,
                         apr_port_t port);

/**
 * Enables a Server for use again
 * @param mc The memcache client object to use
 * @param ms Server to Activate
 */
APR_DECLARE(apr_status_t)
apr_memcache_enable_server(apr_memcache_t *mc,
                           apr_memcache_server_t *ms);


/**
 * Disable a Server
 * @param mc The memcache client object to use
 * @param ms Server to Disable
 */
APR_DECLARE(apr_status_t)
apr_memcache_disable_server(apr_memcache_t *mc,
                            apr_memcache_server_t *ms);

/**
 * Creates a new Server Object
 * @param p Pool to use
 * @param host hostname of the server
 * @param port port of the server
 * @param min  minimum number of client sockets to open
 * @param smax soft maximum number of client connections to open
 * @param max  hard maximum number of client connections
 * @param ttl  time to live in seconds of a client connection
 * @param ns   location of the new server object
 * @see apr_reslist_create
 * @remark min, smax, and max are only used when APR_HAS_THREADS
 */
APR_DECLARE(apr_status_t)
apr_memcache_server_create(apr_pool_t *p,
                           const char *host,
                           apr_port_t port,
                           apr_uint32_t min,
                           apr_uint32_t smax,
                           apr_uint32_t max,
                           apr_uint32_t ttl,
                           apr_memcache_server_t **ns);
/**
 * Creates a new memcached client object
 * @param p Pool to use
 * @param max_servers maximum number of servers
 * @param flags Not currently used
 * @param mc   location of the new memcache client object
 */
APR_DECLARE(apr_status_t) 
apr_memcache_create(apr_pool_t *p,
                    apr_uint16_t max_servers,
                    apr_uint32_t flags,
                    apr_memcache_t **mc);

/**
 * Gets a value from the server, allocating the value out of p
 * @param mc client to use
 * @param p Pool to use
 * @param key null terminated string containing the key
 * @param baton location of the allocated value
 * @param len   length of data at baton
 * @param flags any flags set by the client for this key
 * @return 
 */
APR_DECLARE(apr_status_t) 
apr_memcache_getp(apr_memcache_t *mc, 
                  apr_pool_t *p,
                  const char* key,
                  char **baton,
                  apr_size_t *len,
                  apr_uint32_t *flags);

/**
 * Sets a value by key on the server
 * @param mc client to use
 * @param key   null terminated string containing the key
 * @param baton data to store on the server
 * @param len   length of data at baton
 * @param timeout time in seconds for the data to live on the server
 * @param flags any flags set by the client for this key
 */
APR_DECLARE(apr_status_t) 
apr_memcache_set(apr_memcache_t *mc,
                 const char* key,
                 char *baton,
                 const apr_uint32_t data_size,
                 apr_uint32_t timeout,
                 apr_uint32_t flags);

/**
 * Adds value by key on the server
 * @param mc client to use
 * @param key   null terminated string containing the key
 * @param baton data to store on the server
 * @param len   length of data at baton
 * @param timeout time for the data to live on the server
 * @param flags any flags set by the client for this key
 * @return APR_SUCCESS if the key was added, APR_EEXIST if the key 
 * already exists on the server.
 */
APR_DECLARE(apr_status_t) 
apr_memcache_add(apr_memcache_t *mc,
                 const char* key,
                 char *baton,
                 const apr_uint32_t data_size,
                 apr_uint32_t timeout,
                 apr_uint32_t flags);

/**
 * Replaces value by key on the server
 * @param mc client to use
 * @param key   null terminated string containing the key
 * @param baton data to store on the server
 * @param len   length of data at baton
 * @param timeout time for the data to live on the server
 * @param flags any flags set by the client for this key
 * @return APR_SUCCESS if the key was added, APR_EEXIST if the key 
 * did not exist on the server.
 */
APR_DECLARE(apr_status_t) 
apr_memcache_replace(apr_memcache_t *mc,
                 const char* key,
                 char *data,
                 const apr_uint32_t data_size,
                 apr_uint32_t timeout,
                 apr_uint32_t flags);
/**
 * Deletes a key from a server
 * @param mc client to use
 * @param key   null terminated string containing the key
 * @param timeout time for the delete to stop other clients from adding
 */
APR_DECLARE(apr_status_t) 
apr_memcache_delete(apr_memcache_t *mc,
                 const char* key,
                 apr_uint32_t timeout);

/**
 * Increments a value
 * @param mc client to use
 * @param key   null terminated string containing the key
 * @param n     number to increment by
 * @param nv    new value after incrmenting
 */
APR_DECLARE(apr_status_t) 
apr_memcache_incr(apr_memcache_t *mc, 
                 const char* key,
                 apr_int32_t n,
                 apr_uint32_t *nv);

/**
 * Decrements a value
 * @param mc client to use
 * @param key   null terminated string containing the key
 * @param n     number to decrement by
 * @param nv    new value after decrementing
 */
APR_DECLARE(apr_status_t) 
apr_memcache_decr(apr_memcache_t *mc, 
                 const char* key,
                 apr_int32_t n,
                 apr_uint32_t *new_value);

/**
 * Query a server's version
 * @param ms    server to query
 * @param p     Pool to allocate answer from
 * @param baton location to store server version string
 * @param len   length of the server version string
 */
APR_DECLARE(apr_status_t)
apr_memcache_version(apr_memcache_server_t *ms,
                  apr_pool_t *p,
                  char **baton);

typedef struct
{
    /** Version string of this server */
    const char* version;
    /** Process id of this server process */
    apr_uint32_t pid;
    /** Number of seconds this server has been running */
    apr_uint32_t uptime;
    /** current UNIX time according to the server */
    apr_time_t time;
    /** Accumulated user time for this process */
    apr_time_t rusage_user;
    /** Accumulated system time for this process */
    apr_time_t rusage_system;
    /** Current number of items stored by the server */
    apr_uint32_t curr_items;
    /** Total number of items stored by this server */
    apr_uint32_t total_items;
    /** Current number of bytes used by this server to store items */
    apr_uint64_t bytes;
    /** Number of open connections */
    apr_uint32_t curr_connections;
    /** Total number of connections opened since the server started running */
    apr_uint32_t total_connections;
    /** Number of connection structures allocated by the server */
    apr_uint32_t connection_structures;
    /** Cumulative number of retrieval requests */
    apr_uint32_t cmd_get;
    /** Cumulative number of storage requests */
    apr_uint32_t cmd_set;
    /** Number of keys that have been requested and found present */
    apr_uint32_t get_hits;
    /** Number of items that have been requested and not found */
    apr_uint32_t get_misses;
    /** Total number of bytes read by this server */
    apr_uint64_t bytes_read;
    /** Total number of bytes sent by this server */
    apr_uint64_t bytes_written;
    /** Number of bytes this server is allowed to use for storage. */
    apr_uint32_t limit_maxbytes;
} apr_memcache_stats_t;

/**
 * Query a server for statistics
 * @param ms    server to query
 * @param p     Pool to allocate answer from
 * @param stats location of the new statistics structure
 */
APR_DECLARE(apr_status_t) 
apr_memcache_stats(apr_memcache_server_t *ms, 
                  apr_pool_t *p,
                  apr_memcache_stats_t **stats);


/** @} */

#ifdef __cplusplus
}
#endif

#endif /* APR_MEMCACHE_H */


mod_ip_count.c
/* $Id: mod_ip_count.c,v 1.2 2008/11/22 09:10:07 proger Exp $ */

/*
 * Copyright 2004 Ian Holsman,
 * Copyright 2008 Ivan Fitenko
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define APR_WANT_STRFUNC
#include <apr_want.h>
#include <apr_strings.h>
#include <apr_time.h>
#include <apr_network_io.h>

#include "httpd.h"
#include "http_config.h"
#include "http_core.h"
#include "http_log.h"
#include "http_protocol.h"
#include "http_request.h" 

#include "apr_memcache.h"
#undef PACKAGE_NAME
#undef PACKAGE_STRING
#undef PACKAGE_TARNAME
#undef PACKAGE_VERSION
#include "mod_ip_count.h"
#define DEFAULT_SERVER "127.0.0.1:11211"
#define DEFURI "/"
#define NOWHITELIST "255.255.255.255"

module AP_MODULE_DECLARE_DATA ip_count_module;

typedef struct {
  char *memcache_servers;
  char *used_uris;
  unsigned int memcache_max_requests;
  unsigned int httpretcode;
  apr_short_interval_time_t memcache_max_time;
  apr_short_interval_time_t memcache_block_time;
  apr_short_interval_time_t rq_latency;
  apr_memcache_t *mc;
  char *whitelisted_ip;
  /* TODO: add config settings for these... */
  /* ... or just get this TODO the hell out of here.*/
  apr_uint32_t min_sockets;
  apr_uint32_t smax_sockets;
  apr_uint32_t max_sockets;
  apr_uint32_t ttl_sockets;
} ip_count_config_rec;

typedef struct {
  int size;	/* container for MaxTime setting */
  int nelts;	/* request counter */
  int posn;	/* positon of current request */
  int lastseenposn; /* position of oldest request */
  int pidaras;	/* blacklist token */
  apr_time_t times[]; /* storage for request times*/
} user_details;


static const char *set_servers(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->memcache_servers = apr_pstrdup( cmd->pool, arg1);
    return NULL;
}

static const char *set_used_uris(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->used_uris = apr_pstrdup( cmd->pool, arg1);
    return NULL;
}

static const char *set_max_time(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->memcache_max_time = apr_time_make( atoi( arg1) ,0);

    return NULL;
}

static const char *set_block_time(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->memcache_block_time = apr_time_make( atoi( arg1) ,0);

    return NULL;
}

static const char *set_max_requests(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->memcache_max_requests = atoi( arg1) - 1;
    return NULL;
}

static const char *set_rq_latency(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->rq_latency = atoi( arg1);
    return NULL;
}


static const char *set_httpretcode(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->httpretcode = atoi( arg1);
    return NULL;
}

static const char *set_whitelisted_ip(cmd_parms *cmd, void *config, const char *arg1 )
{
    ip_count_config_rec *cfg;
    cfg = ap_get_module_config(cmd->server->module_config, &ip_count_module );
    cfg->whitelisted_ip = apr_pstrdup( cmd->pool, arg1);
    return NULL;
}

static void *create_ip_count_server_config(apr_pool_t *p, server_rec *s)
{
    ip_count_config_rec *conf = apr_palloc(p, sizeof(*conf));

    conf->memcache_servers = DEFAULT_SERVER;
    conf->used_uris = DEFURI;
    conf->memcache_max_requests = 50;
    conf->memcache_max_time = apr_time_make(120,0);
    conf->memcache_block_time = apr_time_make(120,0);
    conf->httpretcode = 403;
    conf->min_sockets=2;
    conf->smax_sockets = 5;
    conf->max_sockets = 10;
    conf->ttl_sockets= 600;
    conf->mc = NULL;
    conf->whitelisted_ip= NOWHITELIST;
    conf->rq_latency = 0;
    return conf;
}
 
static void ip_count_child_init(apr_pool_t *p, server_rec *s)
{
    char *serverstring;
    char *serverport;
    char *last;
    int nservers=0;

    apr_status_t rv;
    apr_memcache_server_t *serverrec;

    ip_count_config_rec *conf = ap_get_module_config(s->module_config,
                                                      &ip_count_module);
    /* Find all the servers in the first run to get a total count */
    serverstring = apr_pstrdup(p, conf->memcache_servers);
    serverport = apr_strtok(serverstring, " ", &last);
    while (serverport) {
        nservers++;
        serverport = apr_strtok(NULL," ", &last);
    }    
    
    rv = apr_memcache_create(p, nservers, 0, &conf->mc);

    if (rv != APR_SUCCESS) {
        ap_log_error(APLOG_MARK, APLOG_CRIT, rv, s,
                  "apr_memcache_create:unable to initialize memcached");
    }
    serverstring = apr_pstrdup(p, conf->memcache_servers);
    serverport = apr_strtok(serverstring, " ", &last);
    while (serverport) {
       char *server;
       char *scope;
       apr_port_t port;

       rv = apr_parse_addr_port( &server, &scope, &port, serverport, p );
       if ( rv != APR_SUCCESS) {
            ap_log_error(APLOG_MARK, APLOG_CRIT, rv, s,
                      "apr_memcache_server_create: unable to parse %s (%s)", serverport, conf->memcache_servers );
       }
       if (server == NULL ) {
           server = DEFAULT_SERVER;
       }
       
       rv = apr_memcache_server_create(p, 
                        server, 
                        port,
                        conf->min_sockets,
                        conf->smax_sockets,
                        conf->max_sockets,
                        conf->ttl_sockets,
                        &serverrec);
        if (rv != APR_SUCCESS) {
            ap_log_error(APLOG_MARK, APLOG_CRIT, rv, s,
                      "apr_memcache_server_create:%s %d", server, port);
	    return DECLINED;
        }
        rv = apr_memcache_add_server(conf->mc, serverrec);
        if (rv != APR_SUCCESS) {
            ap_log_error(APLOG_MARK, APLOG_CRIT, rv, s,
                      "apr_memcache_add_server:");
        }

       serverport = apr_strtok(NULL," ", &last);
    }
}
/* Checking ID */
static int ip_count_check_auth(request_rec *r)
{
    ip_count_config_rec *conf = ap_get_module_config(r->server->module_config,
                                                      &ip_count_module);
    char *reason = NULL;
    apr_status_t rv;
    user_details *det;
    int i;
    int t_var;
    apr_time_t oldest_time;
    request_rec *rr;
    char datetimestr[APR_RFC822_DATE_LEN];
    apr_time_t *last_seen;
    char *key;
    apr_memcache_t *mc;
    int new=0;
    apr_size_t len;
    apr_uint32_t flags;
    char *result;
    char *uri_samples;
    char* req_uris;
    char* all_whitelisted_ip;
    char* this_whitelisted_ip;
    char* whitelisted_netmask;
    apr_ipsubnet_t* this_subnet;
    char *last;
    
    if (r->main) {
        return DECLINED;
    }
#ifdef WITH_APACHE24
    key = r->connection->client_ip;
#else
    key = r->connection->remote_ip;
#endif
/*    Use this for debugging
 *    key = r->args;
 */

    if (!key) {
        key = "none";
    }
    mc = conf->mc;

    rv = apr_memcache_getp( mc, r->pool, key, &result, &len, &flags );
    if (rv != APR_SUCCESS && rv != APR_NOTFOUND) {
        ap_log_rerror(APLOG_MARK, APLOG_WARNING, rv, r,
                  "Getting cache key for IP %s", key);
        return DECLINED;
    }

    if (rv == APR_NOTFOUND) {
        det = apr_palloc( r->pool, sizeof(user_details) + sizeof(apr_time_t)* conf->memcache_max_requests );
        det->size = conf->memcache_max_requests;
        det->nelts = 0;
	det->pidaras = 0;
        det->posn = 0;
        det->lastseenposn = 0;
        new = 1;
    }
    else {
        det = (user_details*)result;
    }


    /*for the configured uris... */
    req_uris = apr_pstrdup(r->pool, conf->used_uris);
    uri_samples = apr_strtok(req_uris, " ",&last);
    do {
	if (!strstr(r->uri,uri_samples)) {
	return DECLINED;
	}
    uri_samples = apr_strtok(NULL, " ",&last);
    } while (uri_samples);

    /*Process the whitelist, skip the requests mathcing the set criteria*/    
    all_whitelisted_ip = apr_pstrdup(r->pool, conf->whitelisted_ip);
    this_whitelisted_ip = apr_strtok(all_whitelisted_ip, " ",&last);
    do {
	if (!strncmp(this_whitelisted_ip,"env=",4)){
	    if (apr_table_get(r->subprocess_env, (this_whitelisted_ip+4))) {
		return DECLINED;
	    }
	}	    
	if ((whitelisted_netmask = strchr(this_whitelisted_ip, '/'))) {
	    *whitelisted_netmask++ = '\0'; 
	    apr_ipsubnet_create(&this_subnet,this_whitelisted_ip,
		whitelisted_netmask,r->pool);
	} else {
	    apr_ipsubnet_create(&this_subnet,this_whitelisted_ip,NULL,r->pool);
	}
#ifdef WITH_APACHE24
	if (apr_ipsubnet_test(this_subnet,r->connection->client_addr)) {
#else
	if (apr_ipsubnet_test(this_subnet,r->connection->remote_addr)) {
#endif
	    return DECLINED;
	}
	this_whitelisted_ip = apr_strtok(NULL, " ",&last);
    } while (this_whitelisted_ip);

    /* Pass the next request if comes fast enough to be considered the 'dup'*/
    if (conf->rq_latency) {			  
        if ((r->request_time - det->times[ det->posn ]) <= conf->rq_latency) {
	    return DECLINED;
	}
    }

    if (det->pidaras) {
	oldest_time = r->request_time - conf->memcache_block_time;
    } else {
	oldest_time = r->request_time - conf->memcache_max_time;
    }
    
    
    /* remove the requests with expired time */
    last_seen = det->times;
    while  ( det->nelts > 0 && last_seen[ det->lastseenposn ] < oldest_time) {
        det->nelts--;
        det->times[ det->lastseenposn ] = -1;
        det->lastseenposn = ( det->lastseenposn + 1 ) % det->size;
    }

    
    /* ...check if we have gone too far... */
    if ( det->nelts >= det->size || det->nelts >= conf->memcache_max_requests) {
        apr_rfc822_date(datetimestr, det->times[det->lastseenposn] );
        /* ...if so, mark for blacklist... */
        if (!det->pidaras) {
            det->pidaras=1;
            if (conf->memcache_block_time < conf->memcache_max_time) {
                rv = apr_memcache_set( mc, 
			    key, 
			    (char*)det, 
			    sizeof(user_details) + (sizeof(apr_time_t)* det->size),
			    apr_time_sec(conf->memcache_max_time)+600,
			    0);
	    } else {
	        rv = apr_memcache_set( mc, 
			    key, 
		    	    (char*)det, 
			    sizeof(user_details) + (sizeof(apr_time_t)* det->size),
			    apr_time_sec(conf->memcache_block_time)+600,
			    0);
	    }
			
		if (rv) {
    		    ap_log_rerror(APLOG_MARK, APLOG_ERR, rv, r,                		    "Setting cache key for IP %s failed", key);
		}
	}
    
    /* ... and send the predefined response. */
    ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
                  "access to %s blocked for IP %s requests started at %s", r->uri, key, datetimestr);
    return conf->httpretcode;
    }
    
    /* now add this one */

    det->nelts++;
    det->posn = ( det->posn + 1 ) % det->size;
    det->times[det->posn ] = r->request_time;

/* 	Debug reporting for each hit 
 *       ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
 *                  "Setting cache key for IP %s OK: seen %d in the last %d seconds", key, det->nelts, apr_time_sec(conf->memcache_max_time));
*/

    /* keep the record around for a bit longer than the max time */
    if (conf->memcache_block_time < conf->memcache_max_time){
        rv = apr_memcache_set( mc, 
		   key, 
		   (char*)det, 
		   sizeof(user_details) + (sizeof(apr_time_t)* det->size),
		   apr_time_sec(conf->memcache_max_time)+600,
		   0);
    } else {
        rv = apr_memcache_set( mc, 
    		   key, 
		   (char*)det, 
		   sizeof(user_details) + (sizeof(apr_time_t)* det->size),
		   apr_time_sec(conf->memcache_block_time)+600,
		   0);
    }
    
    if (rv) {
    ap_log_rerror(APLOG_MARK, APLOG_ERR, rv, r,
                  "Setting cache key for IP %s failed", key);
    }

    return DECLINED;
}


static const command_rec ip_count_cmds[] =
{
    AP_INIT_TAKE1("MemCacheServers", set_servers, NULL, RSRC_CONF, 
            "a list of servers running the memcached server (host:port)"),
    AP_INIT_TAKE1("UriList", set_used_uris, NULL, RSRC_CONF, 
            "a list of patterns to match for the uris to operate on"),
    AP_INIT_TAKE1("MemCacheMaxRequests", set_max_requests, NULL, RSRC_CONF, 
            "Max number of requests before failing"),
    AP_INIT_TAKE1("MemCacheMaxTime", set_max_time, NULL, RSRC_CONF,
            "Time period in which the requests have to come (seconds)"),
    AP_INIT_TAKE1("MemCacheBlockTime", set_block_time, NULL, RSRC_CONF,
            "Additional blocking time (seconds)"),
    AP_INIT_TAKE1("Latency", set_rq_latency, NULL, RSRC_CONF,
            "time between requests to consider them a single hit (microseconds)"),
    AP_INIT_TAKE1("HttpResponse", set_httpretcode, NULL, RSRC_CONF,
            "An HTTP code to return in response to a blocked request"),	    
    AP_INIT_TAKE1("MemCacheAllow", set_whitelisted_ip, NULL, RSRC_CONF,
            "IP adresses NOT to perform checks on"),	    
    {NULL}
};

static void register_hooks(apr_pool_t *p)
{

    ap_hook_child_init(ip_count_child_init, NULL, NULL, APR_HOOK_MIDDLE);
    ap_hook_fixups(ip_count_check_auth, NULL, NULL, APR_HOOK_MIDDLE);
}

module AP_MODULE_DECLARE_DATA ip_count_module =
{
    STANDARD20_MODULE_STUFF,
    NULL,                         /* dir config creater */
    NULL,                         /* dir merger --- default is to override */
    create_ip_count_server_config,/* server config */
    NULL,                         /* merge server config */
    ip_count_cmds,                /* command apr_table_t */
    register_hooks                /* register hooks */
};
